File: engine/passes/parser-pass.ts
/**
 * ParserPass — CLAUDE.md v2 §8 (first stage of the pipeline)
 *
 * Algorithm: recursive-descent parsing, one function per grammar
 * production. This directly mirrors the "predictive parser" technique
 * from Aho/Lam/Sethi/Ullman (the prescribed BCSE307L textbook), applied
 * to the grammar in types.ts.
 *
 * Design choices worth being able to defend in a viva:
 *   - '|' binds loosest, then concatenation, then '*' — enforced purely
 *     by which grammar production calls which (parseRegex calls
 *     parseTerm calls parseFactor calls parseAtom), NOT by a precedence
 *     table. This is the textbook way recursive descent encodes
 *     precedence: tighter-binding operators live deeper in the call chain.
 *   - Both '|' (alt) and concatenation are built LEFT-ASSOCIATIVE here
 *     (a loop, not right-recursion) — e.g. "a|b|c" becomes
 *     alt(alt(a,b),c). This doesn't change the language accepted
 *     (both operations are associative for regular languages) but it's
 *     a concrete implementation choice you should be able to name.
 *   - Errors are thrown internally as ParseFailure and caught once at
 *     the top of parse(), so every recursive helper can stay simple and
 *     just "throw on trouble" instead of threading Result types through
 *     every call.
 *
 * ONE ADDITION BEYOND THE WRITTEN GRAMMAR: backslash-escaping (e.g. `\(`
 * matches a literal '('). The grammar's CHAR production doesn't exclude
 * this, but it's not spelled out either — flagging it explicitly per
 * CLAUDE.md v2 §0.1 rule 5. It's cheap (a few lines) and lets you write
 * test patterns that contain literal metacharacters. Remove it if you'd
 * rather stay strictly literal to the grammar as written.
 */

import type { ASTNode, ParseError, ParseResult } from "../types.js";

class ParseFailure extends Error {
  constructor(public readonly parseError: ParseError) {
    super(`${parseError.type} at position ${parseError.position}`);
  }
}

class Parser {
  private pos = 0;

  constructor(private readonly pattern: string) {}

  private peek(): string | undefined {
    return this.pattern[this.pos];
  }

  private advance(): string {
    return this.pattern[this.pos++];
  }

  /** regex ::= term ('|' term)* */
  private parseRegex(): ASTNode {
    let node = this.parseTerm();
    while (this.peek() === "|") {
      this.advance();
      const right = this.parseTerm();
      node = { kind: "alt", left: node, right };
    }
    return node;
  }

  /** term ::= factor+  (one or more factors, concatenated) */
  private parseTerm(): ASTNode {
    let node = this.parseFactor();
    while (this.canStartFactor()) {
      const right = this.parseFactor();
      node = { kind: "concat", left: node, right };
    }
    return node;
  }

  /** Lookahead helper: could the next token legally start a new factor? */
  private canStartFactor(): boolean {
    const c = this.peek();
    return c !== undefined && c !== "|" && c !== ")";
  }

  /** factor ::= atom quantifier?   (quantifier restricted to '*' in Core scope) */
  private parseFactor(): ASTNode {
    const atom = this.parseAtom();
    if (this.peek() === "*") {
      this.advance();
      return { kind: "star", child: atom };
    }
    return atom;
  }

  /**
   * atom ::= CHAR | '(' regex ')'
   * ('[' charclass ']' is Stretch B — not implemented yet, see types.ts)
   */
  private parseAtom(): ASTNode {
    const c = this.peek();

    if (c === undefined) {
      throw new ParseFailure({
        type: "UnexpectedToken",
        position: this.pos,
        expected: "a character or '('",
        found: "end of input",
      });
    }

    if (c === "(") {
      const openParenPos = this.pos;
      this.advance();

      if (this.peek() === ")") {
        throw new ParseFailure({
          type: "EmptyGroup",
          position: this.pos,
          expected: "a pattern before ')'",
          found: ")",
        });
      }

      const inner = this.parseRegex();

      if (this.peek() !== ")") {
        throw new ParseFailure({
          type: "UnmatchedParen",
          position: openParenPos,
          expected: "')'",
          found: this.peek() ?? "end of input",
        });
      }
      this.advance(); // consume ')'
      return inner;
    }

    if (c === "|" || c === "*" || c === ")") {
      // An operator appearing where an atom was expected has nothing to
      // apply to — e.g. "*ab", "a||b", "a**" (second '*'), "()" handled above.
      throw new ParseFailure({
        type: "DanglingOperator",
        position: this.pos,
        expected: "a character or '('",
        found: c,
      });
    }

    if (c === "\\") {
      const escapePos = this.pos;
      this.advance();
      const escaped = this.peek();
      if (escaped === undefined) {
        throw new ParseFailure({
          type: "UnexpectedToken",
          position: escapePos,
          expected: "a character after '\\'",
          found: "end of input",
        });
      }
      this.advance();
      return { kind: "char", value: escaped };
    }

    this.advance();
    return { kind: "char", value: c };
  }

  parse(): ParseResult {
    try {
      const ast = this.parseRegex();

      if (this.pos < this.pattern.length) {
        const c = this.pattern[this.pos];
        if (c === ")") {
          throw new ParseFailure({
            type: "UnmatchedParen",
            position: this.pos,
            expected: "end of input",
            found: ")",
          });
        }
        throw new ParseFailure({
          type: "UnexpectedToken",
          position: this.pos,
          expected: "end of input",
          found: c,
        });
      }

      return { ok: true, ast };
    } catch (e) {
      if (e instanceof ParseFailure) {
        return { ok: false, error: e.parseError };
      }
      throw e;
    }
  }
}

/** ParserPass entry point: string in, ParseResult out. Pure, no side effects. */
export function parse(pattern: string): ParseResult {
  return new Parser(pattern).parse();
}

File: engine/passes/parser-pass.cpp
// engine/passes/parser-pass.cpp
//
// Direct C++ port of parser-pass.ts. Same grammar, same recursive-descent
// structure, same precedence-via-call-chain, same left-associative
// alt/concat construction, same backslash-escape addition. Only the
// language changed — see parser-pass.ts's own header comment for the
// full design rationale (still applies unchanged).
//
// TS used exceptions (ParseFailure) caught once at the top of parse();
// this port keeps that exact structure, since it's the cleanest way to
// let every recursive helper "just throw on trouble" in C++ too.

#include "regexlab/passes/parser-pass.hpp"

#include <exception>

namespace regexlab {

namespace {

class ParseFailure : public std::exception {
public:
    explicit ParseFailure(ParseError error) : error_(std::move(error)) {}
    const ParseError& error() const { return error_; }
    const char* what() const noexcept override { return "parse failure"; }

private:
    ParseError error_;
};

class Parser {
public:
    explicit Parser(const std::string& pattern) : pattern_(pattern) {}

    ParseResult parse() {
        try {
            ASTNodePtr ast = parseRegex();

            if (pos_ < pattern_.size()) {
                char c = pattern_[pos_];
                if (c == ')') {
                    throw ParseFailure(ParseError{ParseErrorType::UnmatchedParen,
                                                   static_cast<int>(pos_),
                                                   std::string("end of input"),
                                                   std::string(1, ')')});
                }
                throw ParseFailure(ParseError{ParseErrorType::UnexpectedToken,
                                               static_cast<int>(pos_),
                                               std::string("end of input"),
                                               std::string(1, c)});
            }

            return ParseResult{true, ast, ParseError{}};
        } catch (const ParseFailure& failure) {
            return ParseResult{false, nullptr, failure.error()};
        }
    }

private:
    const std::string& pattern_;
    size_t pos_ = 0;

    std::optional<char> peek() const {
        if (pos_ < pattern_.size()) return pattern_[pos_];
        return std::nullopt;
    }

    char advance() { return pattern_[pos_++]; }

    /** regex ::= term ('|' term)* */
    ASTNodePtr parseRegex() {
        ASTNodePtr node = parseTerm();
        while (peek() == '|') {
            advance();
            ASTNodePtr right = parseTerm();
            node = makeAlt(node, right);
        }
        return node;
    }

    /** term ::= factor+  (one or more factors, concatenated) */
    ASTNodePtr parseTerm() {
        ASTNodePtr node = parseFactor();
        while (canStartFactor()) {
            ASTNodePtr right = parseFactor();
            node = makeConcat(node, right);
        }
        return node;
    }

    /** Lookahead helper: could the next token legally start a new factor? */
    bool canStartFactor() const {
        auto c = peek();
        return c.has_value() && *c != '|' && *c != ')';
    }

    /** factor ::= atom quantifier?   (quantifier restricted to '*' in Core scope) */
    ASTNodePtr parseFactor() {
        ASTNodePtr atom = parseAtom();
        if (peek() == '*') {
            advance();
            return makeStar(atom);
        }
        return atom;
    }

    /**
     * atom ::= CHAR | '(' regex ')'
     * ('[' charclass ']' is Stretch B — not implemented yet, see types.hpp)
     */
    ASTNodePtr parseAtom() {
        auto c = peek();

        if (!c.has_value()) {
            throw ParseFailure(ParseError{ParseErrorType::UnexpectedToken,
                                           static_cast<int>(pos_),
                                           std::string("a character or '('"),
                                           std::string("end of input")});
        }

        if (*c == '(') {
            size_t openParenPos = pos_;
            advance();

            if (peek() == ')') {
                throw ParseFailure(ParseError{ParseErrorType::EmptyGroup,
                                               static_cast<int>(pos_),
                                               std::string("a pattern before ')'"),
                                               std::string(")")});
            }

            ASTNodePtr inner = parseRegex();

            if (peek() != ')') {
                std::string found =
                    peek().has_value() ? std::string(1, *peek()) : std::string("end of input");
                throw ParseFailure(ParseError{ParseErrorType::UnmatchedParen,
                                               static_cast<int>(openParenPos),
                                               std::string("')'"), found});
            }
            advance();  // consume ')'
            return inner;
        }

        if (*c == '|' || *c == '*' || *c == ')') {
            // An operator appearing where an atom was expected has nothing to
            // apply to — e.g. "*ab", "a||b", "a**" (second '*'), "()" handled above.
            throw ParseFailure(ParseError{ParseErrorType::DanglingOperator,
                                           static_cast<int>(pos_),
                                           std::string("a character or '('"),
                                           std::string(1, *c)});
        }

        if (*c == '\\') {
            size_t escapePos = pos_;
            advance();
            auto escaped = peek();
            if (!escaped.has_value()) {
                throw ParseFailure(ParseError{ParseErrorType::UnexpectedToken,
                                               static_cast<int>(escapePos),
                                               std::string("a character after '\\'"),
                                               std::string("end of input")});
            }
            advance();
            return makeChar(*escaped);
        }

        advance();
        return makeChar(*c);
    }
};

}  // namespace

ParseResult parse(const std::string& pattern) {
    Parser parser(pattern);
    return parser.parse();
}

}  // namespace regexlab
